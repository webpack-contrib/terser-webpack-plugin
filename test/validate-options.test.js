import TerserPlugin from '../src';

it('validation', () => {
  /* eslint-disable no-new */
  expect(() => {
    new TerserPlugin({ test: /foo/ });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ test: 'foo' });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ test: [/foo/] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ test: [/foo/, /bar/] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ test: ['foo', 'bar'] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ test: [/foo/, 'bar'] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ test: true });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ test: [true] });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ include: /foo/ });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ include: 'foo' });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ include: [/foo/] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ include: [/foo/, /bar/] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ include: ['foo', 'bar'] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ include: [/foo/, 'bar'] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ include: true });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ include: [true] });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ exclude: /foo/ });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ exclude: 'foo' });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ exclude: [/foo/] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ exclude: [/foo/, /bar/] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ exclude: ['foo', 'bar'] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ exclude: [/foo/, 'bar'] });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ exclude: true });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ exclude: [true] });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ chunkFilter: () => {} });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ chunkFilter: true });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ cache: true });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ cache: false });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ cache: 'path/to/cache/directory' });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ cache: {} });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ cacheKeys() {} });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ cacheKeys: 'test' });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ parallel: true });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ parallel: false });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ parallel: 2 });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ parallel: '2' });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ parallel: {} });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ sourceMap: true });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ sourceMap: false });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ sourceMap: 'true' });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ minify() {} });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ minify: true });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ terserOptions: {} });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ terserOptions: null });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({
      terserOptions: {
        // eslint-disable-next-line no-undefined
        ecma: undefined,
        warnings: false,
        parse: {},
        compress: {},
        mangle: true,
        module: false,
        output: null,
        toplevel: false,
        nameCache: null,
        ie8: false,
        keep_classnames: false,
        keep_fnames: false,
        safari10: false,
      },
    });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ terserOptions: { emca: 5 } });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ extractComments: true });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ extractComments: false });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ extractComments: 'comment' });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ extractComments: /comment/ });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ extractComments() {} });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        condition: true,
      },
    });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        condition: 'comment',
      },
    });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        condition: /comment/,
      },
    });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        condition() {},
      },
    });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        condition: {},
      },
    });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        filename: 'test.js',
      },
    });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        filename() {},
      },
    });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        filename: true,
      },
    });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        banner: true,
      },
    });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        banner: 'banner',
      },
    });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        banner() {},
      },
    });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({
      extractComments: {
        banner: /test/,
      },
    });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ extractComments: { unknown: true } });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ warningsFilter() {} });
  }).not.toThrow();

  expect(() => {
    new TerserPlugin({ warningsFilter: true });
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    new TerserPlugin({ unknown: true });
  }).toThrowErrorMatchingSnapshot();
  /* eslint-enable no-new */
});
