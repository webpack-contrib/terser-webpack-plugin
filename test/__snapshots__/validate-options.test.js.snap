// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation 1`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.test should be one of these:
   [RegExp | non-empty string, ...] | RegExp | non-empty string
   -> Filtering rules.
   Details:
    * options.test should be an array:
      [RegExp | non-empty string, ...]
    * options.test should be one of these:
      RegExp | non-empty string
      -> Filtering rule as regex or string.
      Details:
       * options.test should be an instance of RegExp.
       * options.test should be a non-empty string."
`;

exports[`validation 2`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.test should be one of these:
   [RegExp | non-empty string, ...] | RegExp | non-empty string
   -> Filtering rules.
   Details:
    * options.test[0] should be one of these:
      RegExp | non-empty string
      -> Filtering rule as regex or string.
      Details:
       * options.test[0] should be an instance of RegExp.
       * options.test[0] should be a non-empty string."
`;

exports[`validation 3`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.include should be one of these:
   [RegExp | non-empty string, ...] | RegExp | non-empty string
   -> Filtering rules.
   Details:
    * options.include should be an array:
      [RegExp | non-empty string, ...]
    * options.include should be one of these:
      RegExp | non-empty string
      -> Filtering rule as regex or string.
      Details:
       * options.include should be an instance of RegExp.
       * options.include should be a non-empty string."
`;

exports[`validation 4`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.include should be one of these:
   [RegExp | non-empty string, ...] | RegExp | non-empty string
   -> Filtering rules.
   Details:
    * options.include[0] should be one of these:
      RegExp | non-empty string
      -> Filtering rule as regex or string.
      Details:
       * options.include[0] should be an instance of RegExp.
       * options.include[0] should be a non-empty string."
`;

exports[`validation 5`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.exclude should be one of these:
   [RegExp | non-empty string, ...] | RegExp | non-empty string
   -> Filtering rules.
   Details:
    * options.exclude should be an array:
      [RegExp | non-empty string, ...]
    * options.exclude should be one of these:
      RegExp | non-empty string
      -> Filtering rule as regex or string.
      Details:
       * options.exclude should be an instance of RegExp.
       * options.exclude should be a non-empty string."
`;

exports[`validation 6`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.exclude should be one of these:
   [RegExp | non-empty string, ...] | RegExp | non-empty string
   -> Filtering rules.
   Details:
    * options.exclude[0] should be one of these:
      RegExp | non-empty string
      -> Filtering rule as regex or string.
      Details:
       * options.exclude[0] should be an instance of RegExp.
       * options.exclude[0] should be a non-empty string."
`;

exports[`validation 7`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.cache should be one of these:
   boolean | string
   -> Enable file caching. Ignored in webpack 5, for webpack 5 please use https://webpack.js.org/configuration/other-options/#cache.
   Details:
    * options.cache should be a boolean.
    * options.cache should be a string."
`;

exports[`validation 8`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.cacheKeys should be an instance of function.
   -> Allows you to override default cache keys. Ignored in webpack 5, for webpack 5 please use https://webpack.js.org/configuration/other-options/#cache."
`;

exports[`validation 9`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.parallel should be one of these:
   boolean | integer
   -> Use multi-process parallel running to improve the build speed.
   Details:
    * options.parallel should be a boolean.
    * options.parallel should be a integer."
`;

exports[`validation 10`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.parallel should be one of these:
   boolean | integer
   -> Use multi-process parallel running to improve the build speed.
   Details:
    * options.parallel should be a boolean.
    * options.parallel should be a integer."
`;

exports[`validation 11`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.sourceMap should be a boolean.
   -> Enables/Disables generation of source maps."
`;

exports[`validation 12`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.minify should be an instance of function.
   -> Allows you to override default minify function."
`;

exports[`validation 13`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.terserOptions should be an object:
   object { â€¦ }
   -> Options for \`terser\`."
`;

exports[`validation 14`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.extractComments should be one of these:
   boolean | string | RegExp | function | object { condition?, filename?, banner? }
   -> Whether comments shall be extracted to a separate file.
   Details:
    * options.extractComments.condition should be one of these:
      boolean | string | RegExp | function
      Details:
       * options.extractComments.condition should be a boolean.
       * options.extractComments.condition should be a string.
       * options.extractComments.condition should be an instance of RegExp.
       * options.extractComments.condition should be an instance of function."
`;

exports[`validation 15`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.extractComments should be one of these:
   boolean | string | RegExp | function | object { condition?, filename?, banner? }
   -> Whether comments shall be extracted to a separate file.
   Details:
    * options.extractComments.filename should be one of these:
      string | function
      Details:
       * options.extractComments.filename should be a string.
       * options.extractComments.filename should be an instance of function."
`;

exports[`validation 16`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.extractComments should be one of these:
   boolean | string | RegExp | function | object { condition?, filename?, banner? }
   -> Whether comments shall be extracted to a separate file.
   Details:
    * options.extractComments.banner should be one of these:
      boolean | string | function
      Details:
       * options.extractComments.banner should be a boolean.
       * options.extractComments.banner should be a string.
       * options.extractComments.banner should be an instance of function."
`;

exports[`validation 17`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options.extractComments has an unknown property 'unknown'. These properties are valid:
   object { condition?, filename?, banner? }"
`;

exports[`validation 18`] = `
"Invalid options object. Terser Plugin has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { test?, include?, exclude?, cache?, cacheKeys?, parallel?, sourceMap?, minify?, terserOptions?, extractComments? }"
`;
